includes/models/model.llms.membership.php

```diff
@@ -251,6 +312,92 @@ public function instructors() {
 		return new LLMS_Post_Instructors( $this );
 	}
 
+	/**
+	 * Retrieve courses associated with the membership
+	 *
+	 * @since [version]
+	 *
+	 * @see LLMS_Membership::get_associated_posts()
+	 *
+	 * @return int[]
+	 */
+	protected function query_associated_courses() {
+
+		$courses = array();
+
+		// Retrieve all access plans with a members-only availability restriction for this membership.
+		foreach ( $this->query_associated_posts( 'llms_access_plan', '_llms_availability', 'members', '_llms_availability_restrictions' ) as $plan_id ) {
+			$plan = llms_get_post( $plan_id );
+			if ( $plan ) {
+				$courses[] = $plan->get( 'product_id' );
+			}
+		}
+
+		// Merge in all the autoenrollment courses from the membership and remove duplicates.
+		$courses = array_unique( array_merge( $courses, $this->get_auto_enroll_courses() ) );
+
+		return $courses;
+
+	}
+
+	/**
+	 * Performs a WPDB query to retrieve posts associated with the membership
+	 *
+	 * @since [version]
+	 *
+	 * @see LLMS_Membesrhip::get_associated_posts()
+	 *
+	 * @param string $post_type     Post type to query for an association with.
+	 * @param string $enabled_key   A meta key name, used to check if the association is enabled for the associated post. For example: "_llms_is_restricted"
+	 * @param string $enabled_value The meta value of the `$enabled_key` when the association is enabled. For example "yes" when checking "_llms_is_restricted"..
+	 * @param string $list_key      The meta key name where associations are stored as a serialized array of WP_Post IDs. For example "_llms_restricted_levels".
+	 * @return int[]
+	 */
+	protected function query_associated_posts( $post_type, $enabled_key, $enabled_value, $list_key ) {
+
+		global $wpdb;
+
+		// See if we have a cached result first.
+		$cache = sprintf( 'membership_%1$d_associated_%2$s', $this->get( 'id' ), $post_type );
+		$found = null;
+		$ids   = wp_cache_get( $cache, '', false, $found );
```

Comment:
we rely on wp to invalidate (clean) the object cache, do you think this can be a problem with persistent object caching? you know object-cache.php drop-in...

Reply:
Of course. Any amount of caching is in danger of being screwed up by drop-ins and caching plugins...

This particular area though caches a complicated and potentially slow query in order to prevent the same query from running multiple times on the same page load which I forsee could potentially happening in future utilizations of this.

I think the fact that a drop-in could make this not work as intended is an acceptable risk...

If we don't cache it someone eventually will tell us were not caching a complicated query that we should be and if we do cache it someone will drop in persistent object caching without thinking about the consequences of doing so...

However... given the fact that all persistent object caches do eventually expire this will cause delays in a new post association showing up on the froentend somewhere but won't really break any existing functionality...

Yea that all causes support but caching in this way exists for a reason and I don't want to shy away from using it because shared webhosts cache things forever all the time and then also blame us for being too resource intensive..

Reply:
Yeah I see, I was more thinking about a way to clear this at shutdown, or using a you know class property as "cache". But yeah I see your point.

Reply:
I don't know how to respond to this...

We're ALLOWED and SHOULD use the WP cache API where possible. Yes, there might be caching issues in the world but that doesn't mean we should start building some kind of crazy cache handler code that we then need to maintain and support in order to accommodate a theoretical issue like this...

I don't know man... That feels really crazy to me

Reply:
I think you're right.
Mine was more something like: "Can we do something with little effort that will avoid us getting support requests about object caching issues? e.g. - different approach - flushing this cache on the specific post metas update."
This was more a general thought and whether it was something to explore, you know, if it was worthwhile.

That's all, fine for me! :)